backend/lambdas/jobs/stats_updater.py

```diff
@@ -30,82 +33,86 @@ def _aggregate_stats(events):
         event_name = event["EventName"]
         event_data = event.get("EventData", {})
         if event_name in ["QuerySucceeded", "QueryFailed"]:
-            stats += Counter(
-                {
-                    "TotalQueryCount": 1,
-                    "TotalQuerySucceededCount": 1
-                    if event_name == "QuerySucceeded"
-                    else 0,
-                    "TotalQueryFailedCount": 1 if event_name == "QueryFailed" else 0,
-                    "TotalQueryScannedInBytes": event_data.get("Statistics", {}).get(
-                        "DataScannedInBytes", 0
-                    ),
-                    "TotalQueryTimeInMillis": event_data.get("Statistics", {}).get(
-                        "EngineExecutionTimeInMillis", 0
-                    ),
-                }
-            )
-        if event_name in [
-            "ObjectUpdated",
-            "ObjectUpdateFailed",
-            "ObjectRollbackFailed",
-        ]:
-            stats += Counter(
-                {
-                    "TotalObjectUpdatedCount": 1
-                    if event_name == "ObjectUpdated"
-                    else 0,
-                    "TotalObjectUpdateFailedCount": 1
-                    if event_name == "ObjectUpdateFailed"
-                    else 0,
-                    "TotalObjectRollbackFailedCount": 1
-                    if event_name == "ObjectRollbackFailed"
-                    else 0,
-                }
-            )
+            stats += Counter({
+                "TotalQueryCount": 1,
+                "TotalQuerySucceededCount": 1 if event_name == "QuerySucceeded" else 0,
+                "TotalQueryFailedCount": 1 if event_name == "QueryFailed" else 0,
+                "TotalQueryScannedInBytes": event_data.get("Statistics", {}).get("DataScannedInBytes", 0),
+                "TotalQueryTimeInMillis": event_data.get("Statistics", {}).get("EngineExecutionTimeInMillis", 0),
+            })
+        if event_name in ["ObjectUpdated", "ObjectUpdateFailed", "ObjectRollbackFailed"]:
+            stats += Counter({
+                "TotalObjectUpdatedCount": 1 if event_name == "ObjectUpdated" else 0,
+                "TotalObjectUpdateFailedCount": 1 if event_name == "ObjectUpdateFailed" else 0,
+                "TotalObjectRollbackFailedCount": 1 if event_name == "ObjectRollbackFailed" else 0,
+            })
+            if event_name == "ObjectUpdateFailed":
```

Comment:
ObjectUpdateFailed is an event triggered for a single object update fail, and it means there was no write - so, nothing to revert as deleting the previous version may mean deleting the only version there.

The revert is done from the task currently - why moving it to lambda? I think I'm a bit lost here.

Reply:
From our experience, there were many cases in which ObjectUpdateFailed event indicated that the file was corrupted, causing both Athena and spark which are querying this dataset to fail.
therefore when ObjectUpdateFailed occurs, we try to revert last version.
revert_last() function makes 2 validations:
    1) checks that there are prior versions for this file.
    2) checks that the previous version is not a 'delete marker'.
only if these two conditions are met the revert happens. 
this way we assure we no file will be deleted.
* as the comment below indicates this flow is only usable when deleteOldVersion flag is set to True.

We encountered some cases in which ObjectUpdateFailed was caused by SIGINT/SIGTERM and the deletion task wasn't able to complete the revert before it was down, therefore we moved it to the stream processor.

Reply:
Ok, just to clarify, the file was corrupted before already, or by the solution during a Deletion Job?

Reply:
caused by the solution

Reply:
Ok, this is pretty interesting, it would be great if we could find out what the issue is and reproduce to create a fix.

Do you have any details about what the corrupted objects look like? For instance, 
* do you still have a full log for the `ObjectUpdateFailed` job event?
* do you know if before/after the objects had similar filesize?
* any correlation to specific partitions or it would happen anywhere?
* could it be related to this fixes we did a couple of weeks ago? #193 #186

Reply:
This is the Athena error message we get:
  ```
Your query has the following error(s):
HIVE_CANNOT_OPEN_SPLIT: Error opening Hive split s3://neura-s3f2-test/testing/neura-emr-raw- 
data/year=2020/month=01/day=20/part-00616-b59d4d11-c5bb-433e-92cb-c78623cfa3f0-c000.snappy.parquet (offset=0, length=2480536): null
```
 
Even further interesting is that this error occurs in two scenarios:
    1) Alternately during the forget phase (for successful jobs, when the job ends it stops occurring)
    2) When there is a corrupted file in the dataset as described in the previous reply.

Therefore, we assume that there is a fraction of time during the file rewrite in which the file is not yet completed but is already a part of the dataset. then, if a forget task stops unexpectedly during this small phase (due to a SIGINT/SIGTERM for instance) 
the file remains corrupted in the dataset and the error keeps occurring.

you can try replicate scenario (1) by starting a deletion job and keep to querying the dataset until this error arrives.
I used:
```select * 
FROM [table_name]
where [match_key] in (...)
```
You can try and replicate scenario (2) by starting a deletion job over a dataset with rather large files (250 MB on our case) and to set fargate's Task memory to the smallest available value- 0.5GB  

I did save some logs locally somewhere, I will try to find them and will let you know.

Reply:
Thanks for the details!

I was able to reproduce the error for the first scenario, and I think I have an explanation. Presto, running under Athena, inherits S3 consistency model and in some scenarios makes multiple calls to S3 to fetch the list of files of a given partition, and then to get the content, for instance getting the metadata, and then making multiple calls to fetch different row groups. It could occur that exactly when we put the new version, there is a timeframe where Presto could get a mix of old and new content. In the scenario of reading the metadata from the old file and then reading the rowgroups from new, the bytes wouldn't match up. This is an example error in my scenario: `HIVE_CANNOT_OPEN_SPLIT: Error opening Hive split s3://matteo-tests/cr_basic/product_category=Apparel/part-00000-495c48e6-96d6-4650-aa65-3c36a3516ddd.c000.snappy.parquet (offset=0, length=33554432): can not read class parquet.format.FileMetaData: null` - immediately after re-running the query, it was fine. Now...if you have a high number of fargate tasks and running a query to all partitions, I guess the probability of this error occurring may increase significantly, because while on a specific moment you may be having an issue with an object, a moment later, while the issue could be mitigated for the current one, you may be having it with another one. That would explain why after the job runs all is good. In general, I don't know if I have a solution to this problem, apart from recommending running the deletion job in a moment where read users wouldn't be impacted, or obviously increase the tasks concurrency to reduce the timeframe where this is issue is most likely to happen. 
Do you think this theory may be valid for the scenario n.1?
In this scenario, I think that I wouldn't probably recommend a rollback as it would be completely ineffective (it would just make Presto to suffer from the same issue another time, without de-facto solving the issue).

In regards of the scenario (2) - data staying corrupted after the job ends, I think it's a separate issue, and I'm still trying to reproduce.

Reply:
Sounds very reasonable, and indeed we don't rollback in scenario (1)
 
I emailed you a specific job summary and overview which might be relevant.